hypothesis-python/src/hypothesis/extra/gufunc.py

```diff
@@ -0,0 +1,451 @@
+# This module uses the numpy parser of the Generalized Universal Function API
+# signatures `_parse_gufunc_signature`, which is only available in
+# numpy>=1.12.0 and therefore requires a bump in the requirements for
+# hypothesis.
+from __future__ import absolute_import, division, print_function
+
+from collections import defaultdict
+
+import numpy as np
+import numpy.lib.function_base as npfb
+
+from hypothesis.errors import InvalidArgument
+from hypothesis.extra.numpy import arrays, order_check
+from hypothesis.internal.validation import (
+    check_valid_bound,
+    check_valid_interval,
+)
+from hypothesis.searchstrategy import SearchStrategy
+from hypothesis.strategies import (
+    builds,
+    composite,
+    fixed_dictionaries,
+    integers,
+    just,
+    tuples,
+)
+
+# Should not ever need to broadcast beyond this, but should be able to set it
+# as high as 32 before breaking assumptions in numpy.
+GLOBAL_DIMS_MAX = 12
+
+# Key used in min_side and max_side to indicate min/max on broadcasted dims,
+# using ``object()`` trick to create unique sentinel.
+BCAST_DIM = object()
+# Value used in default dict for max side if variable not specified
+DEFAULT_MAX_SIDE = 5
+
+# This uses "private" function of numpy, but it does the job. It throws a
+# pretty readable exception for invalid input, so we don't need to add anything
+# there. We should also check signature.isascii() since there are lot of weird
+# corner cases with unicode parsing, but isascii() restricts us to Py >=3.7.
+parse_gufunc_signature = npfb._parse_gufunc_signature
+
+
+def check_set_like(arg, name=""):
+    """Validate input can be searched like a `set`."""
+    try:
+        0 in arg
+    except TypeError:
+        raise InvalidArgument("Expected set-like but got %s=%r (type=%s)"
+                              % (name, arg, type(arg).__name__))
+
+
+def check_valid_size_interval(min_size, max_size, name, floor=0):
+    """Check valid for integers strategy and array shapes."""
+    # same checks as done in integers
+    check_valid_bound(min_size, name)
+    check_valid_bound(max_size, name)
+    order_check(name, floor, min_size, max_size)  # ensure non-none & above 0
+    # this is also done in integers, so check for good measure
+    check_valid_interval(min_size, max_size, 'min_size', 'max_size')
+
+
+def order_check_min_max(min_dict, max_dict):
+    """Check min and max dict compatible with integers and array shapes."""
+    check_valid_size_interval(min_dict.default_factory(),
+                              max_dict.default_factory(), "side default")
+    for kk in (set(min_dict.keys()) | set(max_dict.keys())):
+        check_valid_size_interval(min_dict[kk], max_dict[kk], "side %s" % kk)
+
+
+def ensure_int(arg, name=""):
```

Comment:
The current code also casts numpy ints (e.g. `np.int32`) to python ints. Is that desirable? I would think in a numpy module we should be able to accept numpy int types.

Reply:
Unfortunately `int64` and `uint64` misbehave in `math.ceil` and `math.floor`, because the result is cast to a float dtype.  That matches the behaviour of Python2 (it was fixed in Python 3), but means that `floor(x) > x` may be True!

We generally priorities consistence within Hypothesis over consistency with external libraries, so I'd stick to Python's native `integer_types` here.